### PR TITLE
Backend/notification/iss194

### PR DESCRIPTION
**Pull request by aTepliak**

### CODE REVIEW

File: backend/__init__.py [hunk 1]
```diff
@@ -12,7 +12,9 @@
 from flask_principal import PermissionDenied
 from jsonschema import FormatChecker, Draft4Validator
 from pymongo import MongoClient, ReturnDocument
-from mongoengine.fields import UUIDField, ListField, StringField, BooleanField
+from mongoengine.fields import UUIDField, ListField, StringField, BooleanField, ObjectId
+from mongoengine.fields import EmbeddedDocumentField, EmbeddedDocument, ListField, \
+    ObjectIdField
```
Comment (henderkes) [inline backend/__init__.py]: please sum this up in a single import with braces ()

File: backend/__init__.py [hunk 2]
```diff
@@ -108,12 +117,13 @@ class User(DB.Document, UserMixin):
...
+    notifications = DB.ListField(EmbeddedDocumentField(Notification), default=[])
 
 
 class EmailConverter(BaseConverter):
-    regex = r"([a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\." +\
-            r"[a-z0-9!#$%&'*+\/=?^_`"r"{|}~-]+)" +\
-            r"*(@|\sat\s)(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?" +\
+    regex = r"([a-z0-9!#$%&'*+\/=?^_`{|}~-]+(?:\." + \
+            r"[a-z0-9!#$%&'*+\/=?^_`"r"{|}~-]+)" + \
+            r"*(@|\sat\s)(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?" + \
```
Comment (henderkes) [inline backend/__init__.py]: please remove the whitespaces, lint (and pep8?) complain otherwise

File: backend/__init__.py [hunk 3]
```diff
@@ -99,6 +101,13 @@ class Role(DB.Document, RoleMixin):
     description = DB.StringField(max_length=255)
 
 
+class Notification(EmbeddedDocument):
+    notification_id = ObjectIdField(default=ObjectId)
```
Comment (henderkes) [inline backend/__init__.py]: missing DB. I think and ObjectId is not defined, is it?

File: backend/__init__.py [hunk 4]
```diff
@@ -125,6 +135,17 @@ class EmailConverter(BaseConverter):
...
+def notify_users(useremail_list, n_description, n_title, n_link):
+    n = Notification(description=n_description, title=n_title, link=n_link)
+    for email in useremail_list:
+        res = g.user_datastore.get_user(email)
+        res.notifications.append(n)
+        res.save()
+
+@app.before_request
+def notification_func():
+    g.notify_users = notify_users
```
Comment (henderkes) [inline backend/__init__.py]: ugh.... but probably the best we can do before the refactoring

File: backend/api/projects.py [hunk 5]
```diff
@@ -6,16 +6,16 @@
 import time
 import json
 import json5
-
+import uuid
 from flask import request, jsonify, make_response, g, Blueprint
```
Comment (henderkes) [inline backend/api/projects.py]: add a newline please

File: backend/api/projects.py [hunk 6]
```diff
@@ -6,16 +6,16 @@
...
 from flask_security import login_required, roles_required
 from pymongo.collection import ReturnDocument
 
+
```
Comment (henderkes) [inline backend/api/projects.py]: and remove this one

File: backend/api/projects.py [hunk 7]
```diff
@@ -6,16 +6,16 @@
...
 from api.helper import uploader
 from api.helper.apiexception import ApiException
 from globals import ADMIN_PERMISSION
 
-
 projects = Blueprint('api_projects', __name__)
```
Comment (henderkes) [inline backend/api/projects.py]: and include this one again

File: backend/api/projects.py [hunk 8]
```diff
@@ -201,7 +201,6 @@ def update_project(project_id):
     except Exception as err:
         raise ApiException(str(err), 500)
 
-
```
Comment (henderkes) [inline backend/api/projects.py]: see above

File: backend/api/projects.py [hunk 9]
```diff
@@ -406,3 +405,43 @@ def delete_comment(project_id, comment_id):
         raise error
     except Exception as err:
         raise ApiException(str(err), 500)
+
+
+@projects.route('/api/projects/<uuid:project_id>/share/<user_mail>', methods=['POST'])
+@login_required
+def share_via_email(user_mail):
```
Comment (henderkes) [inline backend/api/projects.py]: project_id is missing as parameter I think

File: backend/api/projects.py [hunk 10]
```diff
@@ -406,3 +405,43 @@ def delete_comment(project_id, comment_id):
...
+    """
+    user = g.user_datastore.get_user(user_mail)
+    if not user:
+        return make_response("Unknown User with Email-address: " + user_mail, 404)
+    res = g.projects.find_one({'_id': project_id})
+    if not res:
+        raise ApiException("Project not found", 404)
+
+    description = res["title"] + "was shared with you"
+    link = "/projects/" + str(project_id)
```
Comment (henderkes) [inline backend/api/projects.py:465]: is this the way the link looks at the frontend? I heard something different today

File: backend/api/projects.py [hunk 11]
```diff
@@ -406,3 +405,43 @@ def delete_comment(project_id, comment_id):
...
+@login_required
+def share_with_users(project_id):
+    """Creates and saves notification object
+
+        Returns:
+            res: Notification object as json
+    """
+    emails_list = request.get_json()
+    res = g.projects.find_one({'_id': project_id})
+    if res is None:
```
Comment (henderkes) [inline backend/api/projects.py]: "if not res" please

File: backend/api/projects.py [hunk 12]
```diff
@@ -406,3 +405,43 @@ def delete_comment(project_id, comment_id):
...
+        raise ApiException("Project not found", 404)
+
+    description = res["title"] + "was shared with you"
+    link = "/projects/" + str(project_id)
```
Comment (henderkes) [inline backend/api/projects.py]: see ^2 above

File: backend/api/users.py [hunk 13]
```diff
@@ -57,13 +57,11 @@ def create_user():
...
-                                     email=user['email'],
-                                     password=encrypt_password(
-                                         user['password']),
-                                     bio=user['bio'],
-                                     roles=[roles])
+        g.user_datastore.create_user(first_name=user["first_name"],
+                                     last_name=user["last_name"],
+                                     email=user["email"],
+                                     password=encrypt_password(user["password"]),
+                                     bio=user["bio"], roles=[role])
```
Comment (henderkes) [inline backend/api/users.py:67]: please revert this, the variable is still called "roles" and not "role".

**Review by henderkes — DISMISSED** [review]

File: backend/__init__.py (hunk 4 above)
Comment (Drakulix) [reply backend/__init__.py]: It's again a circular import, there is actual nothing else we can do about it.

File: backend/api/projects.py (hunk 10 above)
Comment (Drakulix) [reply backend/api/projects.py:465]: It might, no idea

**Comment by Drakulix** [comment]
@DubbleClick Probably won't have much more time for this. Would be nice if you could test and finish this, did some final touches.

File: backend/api/users.py (hunk 13 above)
Comment (Drakulix) [reply backend/api/users.py:67]: Changed the variable name instead.

File: backend/__init__.py (hunk 1 above)
Comment (aTepliak) [reply backend/__init__.py]: okay, hope with line breake \ is also ok

File: backend/__init__.py (hunk 3 above)
Comment (aTepliak) [reply backend/__init__.py]: no it is imported

File: backend/__init__.py (hunk 2 above)
Comment (aTepliak) [reply backend/__init__.py]: have no idea, pycharm did it automatically with reformat code i guess. But anyway pep8 didn't complain

File: backend/api/projects.py (hunk 5 above)
Comment (aTepliak) [reply backend/api/projects.py]: ok

File: backend/__init__.py (hunk 1 above)
Comment (henderkes) [reply backend/__init__.py]: It's still two imports from mongoengine.fields now. Also pep8 recommends braces, so 

```
from mongoengine.fields import (UUIDField, ListField, StringField, BooleanField, ObjectId
EmbeddedDocumentField, EmbeddedDocument, ListField, ReferenceField)
```

File: backend/__init__.py [hunk 14]
```diff
@@ -98,6 +99,13 @@ class Role(DB.Document, RoleMixin):
     description = DB.StringField(max_length=255)
 
 
+class Notification(EmbeddedDocument):
```
Comment (henderkes) [inline backend/__init__.py]: make this DB.EmbeddedDocument please, might throw some errors otherwise

File: backend/__init__.py [hunk 15]
```diff
@@ -98,6 +99,13 @@ class Role(DB.Document, RoleMixin):
...
+    notification_id = DB.ObjectIdField(default=ObjectId)
```
Comment (henderkes) [inline backend/__init__.py]: shouldnt this be 
```notification_id = DB.ObjectIdField(default=ObjectId())``` <- edit: no, correct the way it is

DB.EmbeddedDocument, too

File: backend/__init__.py [hunk 16]
```diff
@@ -134,6 +143,23 @@ class EmailConverter(BaseConverter):
 USER_DATASTORE = MongoEngineUserDatastore(DB, User, Role)
 SECURITY = Security(app, USER_DATASTORE)
 
+# internal function to append notifications to the given userlist
+
+
```
Comment (henderkes) [inline backend/__init__.py]: -- please (newlines)

File: backend/api/projects.py [hunk 17]
```diff
@@ -6,6 +6,7 @@
 import uuid
 import json
 import json5
+import uuid
```
Comment (henderkes) [inline backend/api/projects.py]: remove this please, its already imported a few lines above

**Review by henderkes — CHANGES_REQUESTED** [review]

File: backend/api/projects.py (hunk 17 above)
Comment (aTepliak) [reply backend/api/projects.py]: okay, but I did't do anything with this file

File: backend/__init__.py (hunk 15 above)
Comment (aTepliak) [reply backend/__init__.py]: okay, I can try

File: backend/api/projects.py (hunk 17 above)
Comment (henderkes) [reply backend/api/projects.py]: this is a change in this branch though, so should be removed

**Review by henderkes — DISMISSED** [review]

**Review by henderkes — APPROVED** [review]

**Pull request merged**